File: src/main/java/io/github/saneea/feature/json/JsonFormatting.java
package io.github.saneea.feature.json;

import java.io.Reader;
import java.io.Writer;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonElement;

import io.github.saneea.Feature;
import io.github.saneea.FeatureContext;

public class JsonFormatting {

	public static class Pretty extends Base {

		@Override
		public String getShortDescription() {
			return "format JSON with indents";
		}

		@Override
		protected Gson gson() {
			return new GsonBuilder()//
					.setPrettyPrinting()//
					.create();
		}
	}

	public static class Line extends Base {

		@Override
		public String getShortDescription() {
			return "format JSON to line";
		}

		@Override
		protected Gson gson() {
			return new GsonBuilder()//
					.create();
		}
	}

	private static abstract class Base implements Feature, Feature.In.Text.Reader, Feature.Out.Text.Writer {

		private Reader in;
		private Writer out;

		protected abstract Gson gson();

		@Override
		public void run(FeatureContext context) throws Exception {
			Gson gson = gson();
			JsonElement jsonElement = gson.fromJson(in, JsonElement.class);
			gson.toJson(jsonElement, out);
		}

		@Override
		public void setIn(Reader in) {
			this.in = in;
		}

		@Override
		public void setOut(Writer out) {
			this.out = out;
		}
	}

}

File: src/main/java/io/github/saneea/dvh/feature/json/JsonFormatting.kt
package io.github.saneea.dvh.feature.json

import com.google.gson.GsonBuilder
import com.google.gson.JsonElement
import io.github.saneea.dvh.Feature
import io.github.saneea.dvh.Feature.Meta
import io.github.saneea.dvh.FeatureContext
import java.io.Reader
import java.io.Writer

class JsonFormatting {

    class Pretty : Base() {
        override val meta get() = Meta("format JSON with indents")

        override val gsonBuilder: GsonBuilder = GsonBuilder().setPrettyPrinting()
    }

    class Line : Base() {
        override val meta get() = Meta("format JSON to line")

        override val gsonBuilder = GsonBuilder()
    }

    abstract class Base :
        Feature,
        Feature.In.Text.Reader,
        Feature.Out.Text.Writer {

        override lateinit var context: FeatureContext
        override lateinit var inTextReader: Reader
        override lateinit var outTextWriter: Writer

        protected abstract val gsonBuilder: GsonBuilder

        override fun run() {
            val gson = gsonBuilder.create()
            val jsonElement = gson.fromJson(inTextReader, JsonElement::class.java)
            gson.toJson(jsonElement, outTextWriter)
        }
    }
}